examples/vision/barlow_twins.py

```diff
@@ -0,0 +1,990 @@
+        then on-diag)
+
+        Take the sum of the first and second parts and then sum them together.
+
+        Arguments:
+            c: A tf.tensor that represents the cross correlation
+            matrix
+
+        Returns:
+            Returns a tf.tensor which represents the cross correlation
+            matrix with its diagonals as zeros
+        """
+
+        # subtracts diagonals by one and squares them(first part)
+        c_diff = tf.pow(tf.linalg.diag_part(c) - 1, 2)
+
+        # takes off diagonal, squares it, multiplies with lambda(second part)
+        off_diag = tf.pow(self.get_off_diag(c), 2) * self.lambda_amt
+
+        # sum first and second parts together
+        loss = tf.reduce_sum(c_diff) + tf.reduce_sum(off_diag)
+
+        return loss
+
+    def normalize(self, output: tf.Tensor) -> tf.Tensor:
+        """normalize method.
+
+        Normalizes the model prediction
+
+        Arguments:
+            output: the model prediction.
+
+        Returns:
+          Returns a normalized version of the model prediction
+        """
+
+        return (output - tf.reduce_mean(output, axis=0)) / tf.math.reduce_std(
+            output, axis=0
+        )
+
+    def cross_corr_matrix(self, z_a_norm: tf.Tensor, z_b_norm: tf.Tensor) -> tf.Tensor:
+        """cross_corr_matrix method.
+
+        Creates a cross correlation matrix from the predictions.
+        It transposes the first prediction and multiplies this with
+        the second, creating a matrix with shape (n_dense_units, n_dense_units).
+        See build_twin() for more info. Then it divides this with the
+        batch size.
+
+        Arguments:
+            z_a_norm: A normalized version of the first prediction
+            z_b_norm: A normalized version of the second prediction
+
+        Returns:
+          Returns a cross correlation matrix
+        """
+        return (tf.transpose(z_a_norm) @ z_b_norm) / self.batch_size
+
+    def call(self, z_a: tf.Tensor, z_b: tf.Tensor) -> tf.Tensor:
+        """call method.
+
+        Makes the cross-correlation loss. Uses the CreateCrossCorr
+        class to make the cross corr matrix, then finds the loss and
+        returns it(see cross_corr_matrix_loss())
+
+        Arguments:
+            z_a: The prediction of the first set of augmented data
+            z_b: the prediction of the second set of augmented data
+
+        Returns:
+            Returns a (rank-0) tf.Tensor that represents the loss.
+        """
+
+        z_a_norm, z_b_norm = self.normalize(z_a), self.normalize(z_b)
+        c = self.cross_corr_matrix(z_a_norm, z_b_norm)
+        loss = self.cross_corr_matrix_loss(c)
+        return loss
+
+
+"""
+# Barlow Twins' Model Architecture
+The model consists of an
+*   Encoder network, which is a resnet-34
+*   Projector network, which creates the model embeddings.
+   * This consists of an MLP with 3 dense-batchnorm-relu layers
+"""
+
+
+class ResNet34:
```

Comment:
@dewball345 
I think it may not be the usual practice to define the model class this way (L728-L797). We should either use model subclassing API or functional API with a conventional modeling approach. So, here we may not need to define any model-class and simply state only the model functions.

Reply:
Thanks @innat, will inherit the tf.keras.Model class. I believe it is better for it to be in a class because of the complexity of the model

Reply:
Hi, currently i'm having a little problem when converting the resnet to tf.keras model. 

I got the error 
```

    in user code:
    
        File "<ipython-input-11-fdae70ad456f>", line 48, in call  *
            x = keras.layers.Conv2D(64, kernel_size=7, strides=2, padding='same')(x)
        File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 67, in error_handler  **
            raise e.with_traceback(filtered_tb) from None
    
        ValueError: tf.function only supports singleton tf.Variables created on the first call. Make sure the tf.Variable is only created once or created outside tf.function. See https://www.tensorflow.org/guide/function#creating_tfvariables for more information.
    
    
    Call arguments received:
      • inputs=tf.Tensor(shape=(512, 32, 32, 3), dtype=float32)
```

Reply:
@dewball345 
Can you give a separate reproducible file?

Reply:
@dewball345 cc @haifeng-jin 
Is it addressed?

Reply:
I will send a reproducible file later, but for now, I think it's okay the way the resnet structure is. However, I do agree it would be better if the resnet was defined as a model and not as a model factory. 

Could you elaborate on what you meant by original paperwork? Do you mean the original paper on resnets?

Reply:
yes, the residual network paper.